# Lista 4, Métodos Numéricos 2 
Javier Alejandro Ovalle Chiquín, 22103  
Ricardo Josué Morales Contreras, 22289

#### Problema 1. Modelo de Prooducción, Periodo único

Solución del modelo con la librería pulp sin restricción de enteros.

In [2]:
import pulp

# Crear el problema
prob = pulp.LpProblem("Produccion_Optima", pulp.LpMaximize)

# Variables continuas
x1 = pulp.LpVariable("Abrigos", lowBound=0)
x2 = pulp.LpVariable("Chamarras", lowBound=0)
x3 = pulp.LpVariable("Pantalones", lowBound=0)
x4 = pulp.LpVariable("Guantes", lowBound=0)
s1 = pulp.LpVariable("Faltante_Abrigos", lowBound=0)
s2 = pulp.LpVariable("Faltante_Chamarras", lowBound=0)
s3 = pulp.LpVariable("Faltante_Pantalones", lowBound=0)
s4 = pulp.LpVariable("Faltante_Guantes", lowBound=0)

# Objetivo
prob += 30*x1 + 40*x2 + 20*x3 + 10*x4 - 15*s1 - 20*s2 - 10*s3 - 8*s4

# Restricciones de demanda
prob += x1 + s1 == 800
prob += x2 + s2 == 750
prob += x3 + s3 == 600
prob += x4 + s4 == 500

# Restricciones de departamentos
prob += 0.30*x1 + 0.30*x2 + 0.25*x3 + 0.15*x4 <= 1000, "Corte"
prob += 0.25*x1 + 0.35*x2 + 0.30*x3 + 0.10*x4 <= 1000, "Aislamiento"
prob += 0.45*x1 + 0.50*x2 + 0.40*x3 + 0.22*x4 <= 1000, "Costura"
prob += 0.15*x1 + 0.15*x2 + 0.10*x3 + 0.05*x4 <= 1000, "Empaque"

# Resolver
prob.solve(pulp.PULP_CBC_CMD(msg=0))

# Resultados
print("Valor objetivo:", pulp.value(prob.objective))
print("x1:", x1.varValue)
print("x2:", x2.varValue)
print("x3:", x3.varValue)
print("x4:", x4.varValue)
print("s1:", s1.varValue)
print("s2:", s2.varValue)
print("s3:", s3.varValue)
print("s4:", s4.varValue)

Valor objetivo: 64625.0
x1: 800.0
x2: 750.0
x3: 387.5
x4: 500.0
s1: 0.0
s2: 0.0
s3: 212.5
s4: 0.0


Estos resultados coinciden exáctamente con los mostrados por simplex de excel

Solución del problema, añadiendo la restricción de que las variables sean enteras:

In [3]:
import pulp

# Crear el problema
prob = pulp.LpProblem("Produccion_Entera", pulp.LpMaximize)

# Variables enteras
x1 = pulp.LpVariable("Abrigos", lowBound=0, cat='Integer')
x2 = pulp.LpVariable("Chamarras", lowBound=0, cat='Integer')
x3 = pulp.LpVariable("Pantalones", lowBound=0, cat='Integer')
x4 = pulp.LpVariable("Guantes", lowBound=0, cat='Integer')
s1 = pulp.LpVariable("Faltante_Abrigos", lowBound=0, cat='Integer')
s2 = pulp.LpVariable("Faltante_Chamarras", lowBound=0, cat='Integer')
s3 = pulp.LpVariable("Faltante_Pantalones", lowBound=0, cat='Integer')
s4 = pulp.LpVariable("Faltante_Guantes", lowBound=0, cat='Integer')

# Objetivo 
prob += 30*x1 + 40*x2 + 20*x3 + 10*x4 - 15*s1 - 20*s2 - 10*s3 - 8*s4

# Restricciones 
prob += x1 + s1 == 800
prob += x2 + s2 == 750
prob += x3 + s3 == 600
prob += x4 + s4 == 500
prob += 0.30*x1 + 0.30*x2 + 0.25*x3 + 0.15*x4 <= 1000, "Corte"
prob += 0.25*x1 + 0.35*x2 + 0.30*x3 + 0.10*x4 <= 1000, "Aislamiento"
prob += 0.45*x1 + 0.50*x2 + 0.40*x3 + 0.22*x4 <= 1000, "Costura"
prob += 0.15*x1 + 0.15*x2 + 0.10*x3 + 0.05*x4 <= 1000, "Empaque"

# Resolver
prob.solve(pulp.PULP_CBC_CMD(msg=0))

# Resultados
print("Valor objetivo:", pulp.value(prob.objective))
print("x1:", x1.varValue)
print("x2:", x2.varValue)
print("x3:", x3.varValue)
print("x4:", x4.varValue)
print("s1:", s1.varValue)
print("s2:", s2.varValue)
print("s3:", s3.varValue)
print("s4:", s4.varValue)

Valor objetivo: 64622.0
x1: 800.0
x2: 750.0
x3: 388.0
x4: 499.0
s1: 0.0
s2: 0.0
s3: 212.0
s4: 1.0


Como se puede observar varian los resultados en x3 que son el número de pantalones producidos, ya que era la solución que no quedaba entera, modificandolo a que sean 388 pantalones y restanto 1 al número de guantes producidos teniendo 212 pantalones faltantes y 1 guante faltante llegando a la solución optima de 64622. 

Lo interesante viene al compararlo con el método simplex de excel, ya que la solución optima fue dejando 387 pantalones y 213 faltantes, llegando a una subsolucion de 64610, la cual es menos a la solución propuesta por pulp. Esto probablemente se debe a las limitantes de las interaciones con excel.

#### Problema 2. Modelo de Producción, periodos múltiples 